### ISSUE TITLE
LSTM output type

### ISSUE THREAD

**Issue by HappyStorm** [2020-DL-Training-Program/Lab3-Image-Caption#2]
Hi 助教您好，

想請問一下關於我們需要實作的 LSTM 的輸出格式，
由於範例 code 所呼叫的 nn.LSTM 傳進去的是 packed sequence,
故 output 至下一層的 dense layer 時也是 packed sequence 的 shape (即非 [B, T, C] or [T, B, C] 的 shape)

因此想請問我們自己實作的 LSTM, 按照 comment 裡的 TO-DO,
流程上應該會先 unpack packed sequence,
再透過 nn.Linear 操作完 LSTM 的運算再輸出

但是這樣的流程下, 最後輸出並不會是原本 sample 直接 run nn.LSTM 的輸出 type (也就是 packed sequence),
導致後續在計算 loss 的時候會有 Error 產生, 那這部份有需要將 LSTM 輸出的結果再轉回 packed sequence 嗎?
因為如果將輸出再轉回去 packed sequence, loss 雖然會下降的跟 sample code 接近, 但 testing 的結果會很奇怪,
不確定是否輸出完又轉回 packed sequence 會讓 output 輸出意義不同, 所以也想詢問這部分是否有需要做其他調整...
(例如調整 target 的 type or shape)

謝謝~

**Comment by zivzone** [comment]
你好，關於這個部分我們可以再看看實作是不是有那裡不太對或者還是有其他地方有問題。
這邊的 unpack 是希望妳們打資料從 packed  中去撈出來，讀取的方式可以參考 Lab3 補充文件
所以關於那一塊寫除來可能會像是
```
is_packed = isinstance(input_seq, PackedSequence)
if is_packed:
	input_seq, batch_sizes = input_seq
	max_batch_size = batch_sizes[0].item()
	time_step = batch_sizes.shape[0]
			
else:
	batch_sizes = None
	max_batch_size = input_seq.shape[0] if self.batch_first else input_seq.shape[1]
	time_step = input_seq.shape[1] if self.batch_first else input_seq.shape[0] 
```
